riklet/src/runtime/function_runtime.rs

```diff
@@ -0,0 +1,198 @@
+use crate::{
+    cli::{config::Configuration, function_config::FnConfiguration},
+    runtime::{network::RuntimeNetwork, RuntimeError},
+    structs::WorkloadDefinition,
+};
+use async_trait::async_trait;
+use curl::easy::Easy;
+use firepilot::{
+    microvm::{BootSource, Config, Drive, MicroVM, NetworkInterface},
+    Firecracker,
+};
+use lz4::Decoder;
+use proto::worker::InstanceScheduling;
+use std::{
+    fs,
+    fs::File,
+    io::{self, Write},
+    path::{Path, PathBuf},
+    thread,
+};
+use tracing::{event, Level};
+
+use super::{network::function_network::FunctionRuntimeNetwork, Runtime, RuntimeManager};
+
+struct FunctionRuntime {
+    function_config: FnConfiguration,
+    file_path: String,
+    network: FunctionRuntimeNetwork,
+    workload_definition: WorkloadDefinition,
+}
+
+#[async_trait]
+impl Runtime for FunctionRuntime {
+    async fn run(&mut self) -> super::Result<()> {
+        self.network
+            .init()
+            .await
+            .map_err(RuntimeError::NetworkError)?;
+
+        event!(Level::INFO, "Function workload detected");
+
+        event!(Level::INFO, "Define network");
+
+        let firecracker = Firecracker::new(firepilot::FirecrackerOptions {
+            command: Some(self.function_config.firecracker_location.clone()),
+            ..Default::default()
+        })
+        .map_err(RuntimeError::FirecrackerError)?;
+
+        event!(Level::DEBUG, "Creating a new MicroVM");
+        let vm = MicroVM::from(Config {
+            boot_source: BootSource {
+                kernel_image_path: self.function_config.kernel_location.clone(),
+                boot_args: Some(format!(
+                    "console=ttyS0 reboot=k nomodules random.trust_cpu=on panic=1 pci=off tsc=reliable i8042.nokbd i8042.noaux ipv6.disable=1 quiet loglevel=0 ip={}::{}:{}::eth0:off",
+                    self.network.firecracker_ip, self.network.tap_ip, self.network.mask_long)
+                ),
+                initrd_path: None,
+            },
+            drives: vec![Drive {
+                drive_id: "rootfs".to_string(),
+                path_on_host: PathBuf::from(self.file_path.clone()),
+                is_read_only: false,
+                is_root_device: true,
+            }],
+            network_interfaces: vec![NetworkInterface {
+                iface_id: "eth0".to_string(),
+                guest_mac: Some("AA:FC:00:00:00:01".to_string()),
+                host_dev_name: self.network.tap_name().map_err(RuntimeError::NetworkError)?,
+            }],
+        });
+
+        event!(Level::DEBUG, "Starting the MicroVM");
+        thread::spawn(move || -> super::Result<()> {
+            event!(Level::INFO, "Function started");
+            firecracker
+                .start(&vm)
+                .map_err(RuntimeError::FirecrackerError)?;
+            Ok(())
```

Comment:
Please handle errors of firecracker, imo we can remove `Result` from the return type of the thread, or just remove the thread to do it synchronously. Firecracker errors here fall into a void